Automatically use secure WebSocket on HTTPS pages

lutovich opened the pull request:
Previously, driver used in browser used secure WebSocket with 'wss' scheme only when encryption was explicitly turned on in the driver config. On HTTP web pages it is perfectly fine to use either plane or secure WebSocket ('ws' and 'wss' respectively). On HTTPS web pages only 'wss' is allowed. So explicit `{encrypted: true}` configuration was always required for drivers used on HTTPS web pages.

This PR makes driver automatically use secure WebSocket when driver is used on a HTTPS web page.

Resolves #255

lutovich committed:
Automatically use secure WebSocket on HTTPS pages

Previously, driver used in browser used secure WebSocket with 'wss'
scheme only when encryption was explicitly turned on in the driver
config. On HTTP web pages it is perfectly fine to use either plane
or secure WebSocket ('ws' and 'wss' respectively). On HTTPS web
pages only 'wss' is allowed. So explicit `{encrypted: true}`
configuration was always required for drivers used on HTTPS web pages.

This commit makes driver automatically use secure WebSocket when driver
is used on a HTTPS web page.

ali-ince commented on src/v1/internal/ch-websocket.js:
```diff
@@ -233,4 +222,46 @@ function asWindowsFriendlyIPv6Address(scheme, parsedUrl) {
+ * @param {function(): string} protocolSupplier - function that detects protocol of the web page.
+ * @return {{scheme: string|null, error: Neo4jError|null}} object containing either scheme or error.
+ */
+function determineWebSocketScheme(config, protocolSupplier) {
```
Based on the information given in _Security Considerations_ section [here](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_client_applications), it may make sense generating warning message about mixing protocols may not work - if encryption is explicitly turned off or on (i.e. https/ws or http/wss).

lutovich replied:
Added a warn print in https://github.com/neo4j/neo4j-javascript-driver/pull/371/commits/c623d8a93962d0633c53c8deb0d9536f918aa883

ali-ince commented on src/v1/internal/ch-websocket.js:
```diff
@@ -233,4 +222,46 @@ function asWindowsFriendlyIPv6Address(scheme, parsedUrl) {
+
+  if (encrypted === true || encrypted === ENCRYPTION_ON) {
+    // encryption explicitly requested in the config
+    if (!trust || trust === 'TRUST_CUSTOM_CA_SIGNED_CERTIFICATES') {
```
This is totally out of scope of this PR, but I find `TRUST_CUSTOM_CA_SIGNED_CERTIFICATES` setting to be confusing and would prefer `TRUST_SYSTEM_CA_SIGNED_CERTIFICATES` as the default. We may consider revising naming / behaviour in upcoming releases?

lutovich replied:
I'll create a card for this

ali-ince reviewed (approved):
LGTM along with a couple of comments.

lutovich committed:
Warn when driver config inconsistent with web page protocol

Encryption can be explicitly turned on or off in the driver config.
Driver can be used on both HTTP and HTTPS web pages. Configured
encryption has to match the security protocol of the web page. Insecure
WebSockets ('ws') have to be used on insecure pages ('http'). Secure
WebSockets ('wss') have to be used on secure pages ('https').
WebSockets might not work in a mixed content environment,
i.e. 'ws' + 'https' or 'wss' + 'http'.

This commit makes driver print a warning when its configuration is
inconsistent with the web page.
